### PR TITLE
Article: Git

### LINKED ISSUE #184: Article: Git
Proposal to add a new article about Git.

Including sources "Understanding Git":
* [Understanding Git — Data Model](https://hackernoon.com/https-medium-com-zspajich-understanding-git-data-model-95eb16cc99f5)
* [Understanding Git — Branching](https://hackernoon.com/understanding-git-branching-2662f5882f9)
* [Understanding Git — Index](https://hackernoon.com/understanding-git-index-4821a0765cf)

### PR DESCRIPTION
**Pull request by ramesht007**
Closes #184 

## If an article related PR

- [x] I have checked that my suggestion follows the Contribution guide

- [x] I have added which issue this PR closes

- [x] I have 0 warnings from Danger.js

- [x] If I have added images, they are in /files/ folder

### CODE REVIEW

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:75]: Please, look at [Problems covered in DXKB](https://developerexperience.io/problems). Add here a short part about Problems Git Helps to Solve. For example, https://developerexperience.io/problems/poor-code-quality seems as a good fit. If there are other problems which you think Git solves, add them as well. (For example, the Lack of cooperation in a team or the Lack of source control. Don't feel forced to use them, they are just suggestions for you.

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:94]: What do you mean by this User interface maturity?

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:98]: From my personal experience, I would also add Not enough understanding of Git commands 😄

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:88]: You could mention Git tools like Gitkraken, GitHub desktop, Fork, and others.

**Review by ridlees — CHANGES_REQUESTED** [review]
Hello @ramesht007, thank you for your PR. I deeply appreciate the effort that went into it, it is a great article. Since you invested your time and energy, I thought I will do the same. Thus, I have read your text multiple times and included my feedback as a tech writer.

I would suggest going through it. Please, don't feel discouraged. Rather, use these points to improve your already great writing.

**Comment by ramesht007** [comment]
hello, @ridlees 
Thank you for the suggetsions.🙂
I will look into them and make changes.
Thanks

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:53]: So you want it be like this:
1-  **blob** object .........................
2-  **tree** object .............................
3 - **commit** object ..................

Comment (ramesht007) [reply practices/understanding_git.md:53]: Thanks, I will make the changes!.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:52]: sorry for the mess, I will keep in mind for next time.😁😅

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:57]: Sure !

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:65]: 👍

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:67]: 👍

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:43]: 👍

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:76]: 👍

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:77]: sure, no problem. 🙂

File: practices/understanding_git.md
Comment (ridlees) [reply practices/understanding_git.md:53]: There are times that lists can lead to better understanding. It is up to you if you use it. I have added this comment just to show you there are multiple approaches and to let you think about them :)

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:75]: Ok, I will look into it

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:94]: Currently Subversion has a wider range of user interface tools than Git. For example there are SVN plugins available for most popular IDEs. There is a Windows Explorer shell extension. There are a number of native Windows and Mac OS X GUI tools available in ready-to-install packages.

Git's primary user interface is through the command line. There are two graphical interfaces: git-gui (distributed with Git) and qgit, which is making great strides towards providing another feature-complete graphical interface. Also gitk, the graphical history browser, can be more than just a fancy log reader. git-gui and gitk usually work out-of-box for common operating systems, and qgit is being ported to Qt4, which improves its portability. There are some user interface tools in development for Git, namely TortoiseGit, a port of TortoiseSVN. There is also Git Extensions, another explorer shell extension

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:98]: hahahahaha

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:88]: can you explain a bit more on this.

Comment (ridlees) [reply practices/understanding_git.md:88]: Of course. There are tools that simplify Git usage. For example, using a tool like Fork allows you to use all `git commands` with Graphical Interface. Another example is GitKraken with cool brach overview https://www.gitkraken.com.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:53]: Thanks, I will go with your idea, as you are more experienced than me. 😅😁

File: practices/understanding_git.md
Comment (ridlees) [reply practices/understanding_git.md:79]: Please, add Problems Git Helps to solve. 

If a beginner comes to Git as Practice, they have an interest in what issues Git solves.

File: practices/understanding_git.md
Comment (ridlees) [reply practices/understanding_git.md:96]: I see now, but the User interface maturity is not a clear term by itself. Also, did you consider the graphical tools I have mentioned in a different comment (Fork, Github Desktop, GitKraken, and others)? 

I would suggest something like "A developer doesn't know how to use the command-line git interface."

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:99]: I am unsure about this. A binary file by its definition (at least the most common) is a non-text file. So, images, songs, videos, and executables all fall under the definition of binary files.  

Is the pitfall that Git is creating additional binary files? If so, then the Pitfall could be something like Lack of branch management

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:60]:
```suggestion
A **branch** is an independent line of development for a particular project with its commits. How does Git does the branching? On the surface, it looks like a list of commits. However, Git only tracks the latest commit on the branch. Knowing the newest change, Git can easily reconstruct the whole commit list. A pointer named **head** which has a reference to the checkout branch (checkout branch tells Git to record all new commits on that branch) and is also used for tracking. 
```

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:79]: Can you share some, as I looked into the article that you gave here as link and added 2 of them. If would be very helpful if you can guide me with the same. 
Thanks.
like what exact problem are you looking to solve.

**Review by ridlees — COMMENTED** [review]
Hi, @ramesht007. Thank you for going through the suggestions. I am sorry that it took me that long, I though I can finish the preview during the weekend but I did not have enough time to fully commit to it.

File: practices/understanding_git.md
Comment (ridlees) [reply practices/understanding_git.md:79]: It is a structural thing :) Problems Git Helps to Solve should have its own "paragraph" and its headline. There add the "problems Git solves" (I have added some ideas, you have some ideas in your text, I think together they are more than enough)

Comment (ramesht007) [reply practices/understanding_git.md:79]: So, I should add these points as well.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:38]: sure I will look into it.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:83]: ohhkkk

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:100]: 👍

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:97]: ok i will remove it.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:99]: Detection and Properties
Subversion can be used with binary files (it is automatically detected; if that detection fails, you have to mark the file binary yourself). Just like Git.

Only that with Git, the default is to interpret the files as binary to begin with. If you _have_ to have CR+LF line endings (even though most modern programs grok the saner LF-only line endings just fine), you have to tell Git so. Git will then autodetect if a file is text (just like Subversion), and act accordingly. Analogous to Subversion, you can correct an erroneous autodetection by setting a git attribute.

Change Tracking
In an earlier version of git [number?] seemingly minor changes to binary files, such as adjusting brightness on an image, could be different enough that Git interprets them as a new file, causing the content history to split. Since Subversion tracks by file, history for such changes is maintained.

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:60]: okk

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:97]: Since Subversion has a single central repository it is possible to specify read and write access controls in a single location and have them be enforced across the entire project.

Git can operate with a central repository workflow. Read and write access can be specified at the central repository.

File: practices/understanding_git.md
Comment (ridlees) [reply practices/understanding_git.md:99]: In this case, you could go for something like _A content history tracking can have errors_ or _A possible loss of the history data in some cases._

Or _Developer updated Git binary files by hand and lost content history._

File: practices/understanding_git.md
Comment (ridlees) [inline practices/understanding_git.md:2]:
```suggestion
slug: git
```

This change is something small. But it can serve as a quick introduction to slugs. Generally, in the SEO (Search Engine Optimization) community is a common practice to make slugs as short as possible.  The idea is that the short and on-point slug is better for both person and search engine (more on this here https://yoast.com/slug/). 

In this case, this great article is about Git. You don't need to add any other parts to slug :)

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:99]: ok cool

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:2]: okk cool

File: practices/understanding_git.md
Comment (ramesht007) [reply practices/understanding_git.md:30]: ok thanks

File: practices/Git.md
Comment (ridlees) [inline practices/Git.md:64]:
```suggestion
According to the Git's point of view, it has mainly three areas: **Working directory, staging area, and the repository**. When we work on a project on our local machine, Git stores the changes in your computer's filesystem which is **working directory**. The changes that are made to the project will remain in the working directory unless we add them to the staging area using the `git add` command. The staging area describes the initial state of the new commit. You can use the staging area to fine-tune the commits by adding or removing until it suits you, after which you can do `git commit` to save it to the directory as an object. For this process, Git doesn't use any directory but on the other hand, it uses **index** which tracks changes.
```

**Review by ridlees — CHANGES_REQUESTED** [review]
These comments are the final touches before my approval. Keep the good work @ramesht007! 🎉

**Review by ridlees — APPROVED** [review]
Hello @ramesht007, sorry for the delay. I am approving the changes. There is a small uncertainty about the "user interface maturity" as I have mentioned before and I asked to remove the emoji. Please, do so. @prokopsimek, could you give this a look?

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:46]:
```suggestion
![Git_understanding](/files/Git_i_dont_know.png)
[Nvie: Git understanding](https://medium.com/hackernoon/https-medium-com-zspajich-understanding-git-data-model-95eb16cc99f5)
```

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:43]: The second headline should be `Why You Might Want the %{name}`. See [Contribution guide](https://github.com/DXHeroes/knowledge-base-content/blob/master/CONTRIBUTING.md).

Please, change the level to `###`.

```suggestion
### How Does Git Work?
```

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:66]: The third headline should be `Why You Might Want the %{name}`. See [Contribution guide](https://github.com/DXHeroes/knowledge-base-content/blob/master/CONTRIBUTING.md).

```suggestion
## Why You Might Want the Git
```

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:78]:
```suggestion
* [Poor Code quality](/problems/poor-code-quality)
```

add link to existing article

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:97]:
```suggestion
## Common Pitfalls of the Git 
```

File: practices/Git.md
Comment (prokopsimek) [inline practices/Git.md:102]:
```suggestion
- Not enough understanding of Git commands
```

Funny 😄 But please, don't have the fun of readers.

**Review by prokopsimek — CHANGES_REQUESTED** [review]
Awesome! There are some minor typos/suggestions.

**Review by prokopsimek — APPROVED** [review]
@ramesht007 Awesome! Thank you for your contribution.  👏🤠

**Pull request merged**
